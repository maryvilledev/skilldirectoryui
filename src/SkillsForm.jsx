import React from 'react';
import axios from 'axios';
import { Col, ControlLabel, Button,
   Form, FormControl, FormGroup } from 'react-bootstrap'

const skillTypes = [
  { value: 'compiled', label: 'Compiled' },
  { value: 'scripted', label: 'Scripted' },
  { value: 'database', label: 'Database' },
  { value: 'orchestration', label: 'Orchestration' }
];

class SkillForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      skill_name: "",
      skill_type: skillTypes[0].value
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(key, value) {
    this.setState({ [key]: value});
  }

  onSubmit(ev) {
    ev.preventDefault();
    const skill_name = this.state.skill_name;
    const skill_type = this.state.skill_type;

    axios.post(this.props.api + '/skills/', {
        name: skill_name,
        skill_type: skill_type
    })
    .then(function (response) {
      console.log(response);
      this.props.closeModal();
    }.bind(this))
    .catch(err => {
      console.log('caught an error', err);
    });
  }

  render() {
    const onSkillNameChange = ev => this.onChange("skill_name", ev.target.value);
    const onSkillTypeChange = ev => this.onChange("skill_type", ev.target.value);
    const options = skillTypes.map(type =>
       (<option key={type.value} value={type.value}>{type.label}</option>)
       );
    return (
      <div>
        <h1>Add Skill</h1>
        <Form horizontal onSubmit={this.onSubmit}>
          <FormGroup controlId='skillName'>
            <Col componentClass={ControlLabel} sm={2}>
              Name:
            </Col>
            <Col sm={10}>
              <FormControl name='skillName' onChange={onSkillNameChange}/>
            </Col>
          </FormGroup>
          <FormGroup controlId='skillType'>
            <Col componentClass={ControlLabel} sm={2}>
              Type:
            </Col>
            <Col sm={10}>
              <FormControl componentClass="select" onChange={onSkillTypeChange}>
                {options}
              </FormControl>
            </Col>
          </FormGroup>
          <FormGroup>
            <Col smOffset={2} sm={10}>
              <Button type="submit" bsStyle="primary">
                Submit
              </Button>
            </Col>
          </FormGroup>
        </Form>
      </div>
    );
  }
}

export default SkillForm
