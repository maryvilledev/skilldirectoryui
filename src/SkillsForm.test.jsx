import React from 'react';
import ReactDOM from 'react-dom';
import SkillsForm from './SkillsForm';
import { shallow } from 'enzyme';
import { mount } from 'enzyme';

describe('<SkillsForm />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SkillsForm />, div);
  });

  it('matches the stored snapshot', () => {
    const wrapper = shallow(<SkillsForm />);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('changes state when input to "Name:" field changes', () => {
    const wrapper = mount(<SkillsForm />);
    // process.stdout.write(wrapper.debug());
    // process.stdout.write(JSON.stringify(wrapper.state()));
    expect(wrapper.state().skill_name).toBe("");
    const event = {target: {value: "New Value"}};
    wrapper.find({ name: "skillName" }).simulate("change", event);
    expect(wrapper.state().skill_name).toBe("New Value");
    // process.stdout.write(JSON.stringify(wrapper.state()));
  });

// TODO: Find different way to test this... The below test fails because
// the "Select" element we are using from 'react-select' doesn't respond
// as expected to the simulated "change" event.
  // it('changes state when selected option in "Skill Type" field changes', () => {
  //   const wrapper = mount(<SkillsForm />);
  //   expect(wrapper.state().skill_type).toBe("");
  //   const event = {target: {value: "New Value"}};
  //   wrapper.find("Select").simulate("change", event);
  //   expect(wrapper.state().skill_type).toBe("New Value");
  // });
});
