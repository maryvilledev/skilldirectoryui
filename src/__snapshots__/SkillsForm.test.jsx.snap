exports[`<SkillsForm /> matches the stored snapshot 1`] = `
<div>
  <h1>
    Add Skill
  </h1>
  <Form
    bsClass="form"
    componentClass="form"
    horizontal={true}
    inline={false}
    onSubmit={[Function]}>
    <FormGroup
      bsClass="form-group"
      controlId="skillName">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={2}>
        Name:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}>
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="skillName"
          onChange={[Function]} />
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
      controlId="skillType">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={2}>
        Type:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}>
        <FormControl
          bsClass="form-control"
          componentClass="select"
          onChange={[Function]}>
          <option
            value="compiled">
            Compiled
          </option>
          <option
            value="scripted">
            Scripted
          </option>
          <option
            value="database">
            Database
          </option>
          <option
            value="orchestration">
            Orchestration
          </option>
        </FormControl>
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group">
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}
        smOffset={2}>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="primary"
          disabled={false}
          type="submit">
          Submit
        </Button>
      </Col>
    </FormGroup>
  </Form>
</div>
`;
